core/src/main/java/org/infinispan/persistence/manager/PersistenceManager.java

```diff
@@ -57,13 +58,46 @@
 
    boolean deleteFromAllStores(Object key, AccessMode mode);
 
-   void processOnAllStores(KeyFilter keyFilter, AdvancedCacheLoader.CacheLoaderTask task, boolean fetchValue, boolean fetchMetadata);
-
-   void processOnAllStores(Executor executor, KeyFilter keyFilter, AdvancedCacheLoader.CacheLoaderTask task, boolean fetchValue, boolean fetchMetadata);
+   /**
+    * See {@link #publishEntries(Predicate, boolean, boolean, AccessMode)}
+    */
+   default <K, V> Publisher<MarshalledEntry<K, V>> publishEntries(boolean fetchValue, boolean fetchMetadata) {
+      return publishEntries(StreamMarshalling.alwaysTruePredicate(), fetchValue, fetchMetadata, AccessMode.BOTH);
+   }
 
-   void processOnAllStores(KeyFilter keyFilter, AdvancedCacheLoader.CacheLoaderTask task, boolean fetchValue, boolean fetchMetadata, AccessMode mode);
+   /**
+    * Returns a publisher that will publish all entries stored by the underlying cache store. Only the first
+    * cache store that implements {@link AdvancedCacheLoader} will be used. Predicate is applied by the underlying
+    * loader in a best attempt to improve performance.
```

Comment:
If it doesn't implement `AdvancedCacheLoader` we can't use it since it doesn't implement the needed method. This is how it used to work before, the big change now is we don't invoke the publisher on _every_ store that implements `AdvancedCacheLoader` which to me was a big bug.

Reply:
Hrmm, I worry about duplicates with doing this, but I can add back in supporting more than 1 loader.

Reply:
I don't think we can put the requirement that you have to implement `CacheWriter` if you implement `AdvancedCacheLoader`. I can see a user possibly wanting a read only backed store that doesn't allow writes. This is a use case of invalidation imo. To be honest `JPAStore` also falls into this niche (it probably should have had a read only and read/write variant tbh).

Also I still really don't like the idea of loading from more than 1 cache loader. We will almost certainly have duplicates, and there is no reasonable way we can limit the memory usage to filter said duplicates out (segments will help but still have issues most likely).

Reply:
Looks like I edited my comment too much and I removed the part where I was agreeing with you that we should only load from one `AdvancedCacheLoader`. I was mentioning `CacheWriter` because 2 loaders that are also writers *should* have the same entries, so there's no benefit in loading entries for both there. But TBH I never understood the use case for using multiple writers on the same cache :)

Anyway, I think restricting publishKeys/Entries to a single `AdvancedCacheLoader` is a good idea, we just need to make the user aware of it.

Reply:
I created https://issues.jboss.org/browse/ISPN-9144